src/core/maprenderer/qgsmaprendererjob.cpp

```diff
@@ -651,66 +680,124 @@ std::vector< LayerRenderJob > QgsMapRendererJob::prepareSecondPassJobs( std::vec
     }
-    if ( layerHasMask.contains( job.layerId ) )
+    else
     {
       // Note: we only need an alpha channel here, rather than a full RGBA image
-      job.context()->setMaskPainter( allocateImageAndPainter( job.layerId, job.maskImage ) );
-      job.maskImage->fill( 0 );
+      QImage *maskImage;
+      labelJob.context.setMaskPainter( allocateImageAndPainter( QStringLiteral( "label mask" ), maskImage ), maskId );
+      maskImage->fill( 0 );
+      labelJob.maskImages.push_back( maskImage );
     }
   }
+  labelJob.context.setMaskIdProvider( &labelJob.maskIdProvider );
+
+  // Prepare second pass jobs
+  // - For raster rendering or vector rendering if effects are involved
+  // 1st pass, 2nd pass and mask are rendered in QImage and composed in composeSecondPass
+  // - For vector rendering if no effects are involved
+  // 1st pass is rendered in QImage, clip paths are generated according to mask and used during
+  // masked symbol layer rendering during second pass, which is rendered in QPicture, second
+  // pass job imgPic
 
   // Allocate an image for labels
-  if ( labelJob.img == nullptr )
+  if ( labelJob.img == nullptr && !forceVector )
   {
     labelJob.img = allocateImage( QStringLiteral( "labels" ) );
   }
+  else if ( labelJob.imgPic == nullptr && forceVector )
+  {
+    labelJob.imgPic.reset( new QPicture() );
+  }
 
-  // Prepare label mask images
-  for ( int maskId = 0; maskId < labelJob.maskIdProvider.size(); maskId++ )
+  // first we initialize painter and mask painter for all jobs
+  for ( LayerRenderJob &job : firstPassJobs )
   {
-    QImage *maskImage;
-    labelJob.context.setMaskPainter( allocateImageAndPainter( QStringLiteral( "label mask" ), maskImage ), maskId );
-    maskImage->fill( 0 );
-    labelJob.maskImages.push_back( maskImage );
+    job.needRasterization = false;
+
+    auto it = maskedSymbolLayers.find( job.layerId );
+    if ( it != maskedSymbolLayers.end() )
+    {
+      QList<MaskSource> &sourceList = it->second;
+      for ( MaskSource &source : sourceList )
+      {
+        job.needRasterization |= source.hasEffects;
+      }
+    }
+
+    // update first pass job painter and device if needed
+    const bool isRasterRendering = !forceVector || job.needRasterization || ( job.renderer && job.renderer->forceRasterRender() );
+    if ( isRasterRendering && !job.img )
+    {
+      job.context()->setPainter( allocateImageAndPainter( job.layerId, job.img ) );
+    }
+    else if ( !isRasterRendering && !job.imgPic )
+    {
+      PictureAndPainter pictureAndPainter = allocatePictureAndPainter();
+      job.imgPic.reset( pictureAndPainter.first );
+      job.context()->setPainter( pictureAndPainter.second );
+      // force recreation of layer renderer so it initialize correctly the renderer
+      // especially the RasterLayerRender that need logicalDpiX from painting device
+      job.renderer = job.layer->createMapRenderer( *( job.context() ) );
```

Comment:
This is already called later in [existing code](https://github.com/qgis/QGIS/blob/master/src/core/maprenderer/qgsmaprendererjob.cpp#L759).

This call is because of [this](https://github.com/troopa81/QGIS/blob/selective_masking_vectorization/src/core/raster/qgsrasterlayerrenderer.cpp#L207), the raster view port is computed in renderer constructor according to given paint device. So if we change the device, we need to recreate the renderer. 

I could update the raster viewport if I detect a dpi change in the render function but it seems a little bit ugly.

Reply:
Sorry - my mistake here too. I mistakenly thought this was getting called on the background thread.